src/zope/formlib/errors.py

```diff
@@ -53,7 +53,7 @@ def snippet(self):
         return u'<span class="error">%s</span>' % escape(msg)
 
 
-@implementer(IBrowserPage)
+
 @adapter(IInvalidCSRFTokenError, IBrowserRequest)
 class InvalidCSRFTokenErrorView(BrowserPage):
```

Comment:
This series of changes makes me a bit sad that the intent of each component (this class implements IFoo) gets removed because it's implied by some superclass we don't see here, and restating it here for clarity makes things break in unobvious ways.

Conclusion: multiple inheritance is evil?

Reply:
The other option I can think of is `@implementer_if_needed()` decorator I've used in a few places, like [zope.schema](https://github.com/zopefoundation/zope.schema/blob/master/src/zope/schema/_field.py#L158). 

I certainly feel that has its uses, but here, the class was directly inheriting from something that pretty clearly (in my opinion) states what it does and is and is unlikely to change, so it didn't seem necessary to me. (There isn't multiple inheritance here, at least not obviously.)

In the `textwidgets.py` case, FloatWidget and DecimalWidget were the outliers. Other things that extend `TextWidget`, like `IntWidget` didn't re-declare the interface implemented by a distant ancestor. (I went back in the history as far back as it exists in this repo, and it has always been that way with no indication why). So opinions seem mixed. 

I'm happy to bring `@implementer_if_needed` to this package if you feel it adds clarity.

Reply:
(This discussion is going meta, and I'm not sure the best place for it is in the comments of a random PR we'll never be able to find later.)

I liked @implementer_if_needed, but it's a shame it has to be copied and pasted everywhere.  Maybe it should live in zope.interface itself?

Maybe we should take it a step further and make @implementer do that always (and have a lower-level decorator with a longer name do what @implementer does today)?

Reply:
> I liked @implementer_if_needed, but it's a shame it has to be copied and pasted everywhere. Maybe it should live in zope.interface itself?

I'm not at all opposed to that if we find it being widely used (I think right now it's in two packages, this would be a third). The main issue is that it would either force those packages to have a hard dependency on a new zope.interface (which forces dropping 3.4) or they'd still have to have a copy of it for backwards compatibility.

> Maybe we should take it a step further and make `@Implementer` do that always (and have a lower-level decorator with a longer name do what `@Implementer` does today)?

That thought occurred to me, but I was very concerned that it would break backwards compatibility.. Right now, in non-strict-IRO mode, one can use a "redundant" `@implementer` declaration to forcibly change the `__iro__`, bringing an inherited interface from the back closer to the front. I don't know how widespread that might be, nor do I know how to attempt to evaluate it or slow-roll a deployment of it. (I haven't really given that last part much thought; it may be possible to do something where we default to strict IRO; if that fails, try implementer-if-needed mode, and if both fail go back to the legacy behaviour. That seems to complicate the test matrix though.)

Reply:
Continued in https://github.com/zopefoundation/zope.interface/issues/199, I suppose.